src/wm/mod.rs

```diff
@@ -0,0 +1,173 @@
+//! Wikimedia types
+use std::{collections::HashSet, ffi::OsStr, fs, num::ParseIntError, str::FromStr};
+
+use anyhow::{anyhow, bail, Context};
+
+use url::Url;
+
+mod page;
+pub use page::Page;
+
+/// Read from a file of urls on each line.
+pub fn parse_wikidata_file(path: impl AsRef<OsStr>) -> anyhow::Result<HashSet<WikidataQid>> {
+    let contents = fs::read_to_string(path.as_ref())?;
+    contents
+        .lines()
+        .enumerate()
+        .map(|(i, line)| {
+            WikidataQid::from_str(line).with_context(|| {
+                let line_num = i + 1;
+                format!("bad QID value on line {line_num}: {line:?}")
+            })
+        })
+        .collect()
+}
+
+/// Read article titles from a file of urls on each line.
+pub fn parse_wikipedia_file(
+    path: impl AsRef<OsStr>,
+) -> anyhow::Result<HashSet<WikipediaTitleNorm>> {
+    let contents = fs::read_to_string(path.as_ref())?;
+    contents
+        .lines()
+        .enumerate()
+        .map(|(i, line)| {
+            WikipediaTitleNorm::from_url(line).with_context(|| {
+                let line_num = i + 1;
+                format!("bad wikipedia url on line {line_num}: {line:?}")
+            })
+        })
+        .collect()
+}
+
+pub fn is_wikidata_match(ids: &HashSet<WikidataQid>, page: &Page) -> Option<WikidataQid> {
+    let Some(wikidata) = &page.main_entity else { return None;};
+    let wikidata_id = &wikidata.identifier;
+    let wikidata_id = match WikidataQid::from_str(wikidata_id) {
+        Ok(qid) => qid,
+        Err(e) => {
+            eprintln!("Could not parse QID: {:?}: {}", wikidata_id, e);
+            return None;
+        }
+    };
+
+    ids.get(&wikidata_id).map(|_| wikidata_id)
+}
+
+pub fn is_wikipedia_match(
+    titles: &HashSet<WikipediaTitleNorm>,
+    page: &Page,
+) -> Option<WikipediaTitleNorm> {
+    // TODO: handle multiple languages
```

Comment:
1. Is it in English only now?
2. Does it make sense to test this function?

Reply:
Oops, that comment slipped through when I handled multiple languages in 6e5385d. I'll remove it.

> 1. Is it in English only now?

Currently it will work with any language, but it only processes a single dump at a time. So when it reads an english dump, each article json has a `.in_language.identifier` field set to `en`, and the program writes that html to a `QXXXXX/en.html` file.

Running the program multiple times with different language dumps will fill in the various `QXXXXX/$lang.html` files.
We could extend it to process multiple dumps in parallel, but I don't expect there to be much of a speedup right now.

> 2. Does it make sense to test this function?

The doctest on the type definition verifies that the two constructors parse and normalize correctly, but doesn't check for various error cases.
Do you think there should be more?

Reply:
1. Does it imply a wrapping script that launches the app for each language/file? That is ok (and can be paralleled in bash by launching several processes simultaneously), but should be documented. Or is there a better approach?
2. It may make sense to check values that are in OSM. Users can make a lot of mistakes.

Reply:
> 1. Does it imply a wrapping script that launches the app for each language/file? That is ok (and can be paralleled in bash by launching several processes simultaneously), but should be documented. Or is there a better approach?

I haven't tried it out yet, but there are a couple of options that come to mind:
- Decompress the archives serially so they're all concatenated together into stdin. This looks possible with `gunzip`/`tar`, not sure about python `pgzip`.
- As you say, run the program repeatedly through a wrapper script, using a for loop, `xargs`, `parallel`, etc.
- Pass the decompression command to the program and have it spawn the subprocess directly, it could do this in parallel and pass the results to the same worker pool.

> 2. It may make sense to check values that are in OSM. Users can make a lot of mistakes.

Understood. I have been using the world list of urls/ids from the map generator with no problems, but if we switch to using OSM data directly I'll rethink this. The program will log any issues it has parsing titles/QIDs.

Reply:
It should be ok to run tasks in parallel using bash for loop. We'll rethink it if necessary later.